# 01. Importing Libraries

In [1]:
#import libraries

import pandas as pd

import numpy as np

import os

# 02. Importing Data

In [8]:
#importing data (ords_prods_merged)
df_ords_prods_merge = pd.read_pickle (r'C:\Users\User\OneDrive\Dokumentumok\CF\Phyton\Instacart Basket Analysis\02 Data\Prepared Data\orders_products_merged.pkl')

# 03. Create a smaller dataframe

In [10]:
#only the first 1 million rows
df = df_ords_prods_merge[:1000000]

# 04. Create new variables

You’ll apply the following criteria:
If the item’s price is lower than or equal to USD 5, it will be labeled a “low-range product.”
If the item’s price is above USD 5 but lower than or equal to USD 15, it will be labeled a “mid-range product.”
If the item’s price is above USD 15, it will be labeled a “high-range product.”

In [11]:
# create a function called price_label
def price_label(row):

  if row['prices'] <= 5:
    return 'Low-range product'
  elif (row['prices'] > 5) and (row['prices'] <= 15):
    return 'Mid-range product'
  elif row['prices'] > 15:
    return 'High range'
  else: return 'Not enough data'

In [ ]:
#the syntax df['price_range'] creates a new column within the df dataframe called “price_range” 
#and designates it as the location for your labels. 
df['price_range'] = df.apply(price_label, axis=1)

In [13]:
df.loc[:, 'price_range'] = df.apply(price_label, axis=1)

In [14]:
#counting how many products are in each categories
df['price_range'].value_counts(dropna = False)

price_range
Mid-range product    652638
Low-range product    338018
High range             9344
Name: count, dtype: int64

In [15]:
#what is the most expensive?
df['prices'].max()

24.5

# 05. If-Statements with the loc() Function

The two implied halves of the if-statement, then, would be:

if = df.loc[df['prices'] > 15,
then = 'price_range_loc'] = 'High-range product'

Remember—the comma is key! It’s what separates the “if” from the “then.”

In [ ]:
df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'

In [19]:
df.loc[(df['prices'] <= 15) & (df['prices'] > 5), 'price_range_loc'] = 'Mid-range product' 

In [20]:
df.loc[df['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [21]:
#counting how many products are in each categories
df['price_range'].value_counts(dropna = False)

price_range
Mid-range product    652638
Low-range product    338018
High range             9344
Name: count, dtype: int64

###  loc() works faster,  repeating the process on the entire dataframe

In [22]:
df_ords_prods_merge.loc[df_ords_prods_merge['prices'] > 15, 'price_range_loc'] = 'High-range product'

In [23]:
df_ords_prods_merge.loc[(df_ords_prods_merge['prices'] <= 15) & (df_ords_prods_merge['prices'] > 5), 'price_range_loc'] = 'Mid-range product' 

In [24]:
df_ords_prods_merge.loc[df_ords_prods_merge['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [26]:
#counting how many products are in each categories
df_ords_prods_merge['price_range_loc'].value_counts(dropna = False)

price_range_loc
Mid-range product     21860868
Low-range product     10126324
High-range product      417678
NaN                      30200
Name: count, dtype: int64

# 06. For-Loop

In [27]:
#Example:
for x in range(30, 45):
    print("My age is %d" % (x))

My age is 30
My age is 31
My age is 32
My age is 33
My age is 34
My age is 35
My age is 36
My age is 37
My age is 38
My age is 39
My age is 40
My age is 41
My age is 42
My age is 43
My age is 44


In [34]:
#on which day most orders take place
#0.0 means Saturday, 0.1 Sunday... 4.0 Wednesday 
#Sunday is busiest, Wednesday is the slowest day
df_ords_prods_merge['orders_day_of_week'].value_counts (dropna=False)

orders_day_of_week
0.0    6209808
1.0    5665951
6.0    4500391
2.0    4217868
5.0    4209603
3.0    3844175
4.0    3787263
NaN         11
Name: count, dtype: int64

You want to use this information to create a new column, “busiest day,”
that will contain one of three different values: 
“Busiest day,” 
“Least busy,” and
“Regularly busy.” 

In [46]:
#The first step is to create an empty list, result. 
#This will act as the empty shell, into which you can place the results from your loop.
result = []

for value in df_ords_prods_merge["orders_day_of_week"]:
  if value == 0:
    result.append("Busiest day")
  elif value == 4:
    result.append("Least busy")
  else:
    result.append("Regularly busy")

In [ ]:
# This value is simply acting as a placeholder. It could stand for anything.
#Here, it represents every possible value within the “orders_day_of_week” column

In [36]:
result

['Regularly busy',
 'Regularly busy',
 'Busiest day',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Busiest day',
 'Busiest day',
 'Busiest day',
 'Busiest day',
 'Busiest day',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Busiest day',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Busiest day',
 'Regularly busy',
 'Busiest day',
 'Least busy',
 'Busiest day',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Bus

### Combine the result list with the dataframe

In [37]:
#create a new column in df and fill it up with elements of result list
df_ords_prods_merge['busiest_day'] = result

In [38]:
# print the frequency of this new column and cross-check it with the frequency of “orders_day_of_week” you printed before.
df_ords_prods_merge['busiest_day'].value_counts (dropna=False)

busiest_day
Regularly busy    22437999
Busiest day        6209808
Least busy         3787263
Name: count, dtype: int64

In [ ]:
#cross-check DONE

# 07. TASKS

#### 02 Suppose your clients have changed their minds about the labels you created in your “busiest_day” column. Now, they want “Busiest day” to become “Busiest days” (plural). This label should correspond with the two busiest days of the week as opposed to the single busiest day. At the same time, they’d also like to know the two slowest days. Create a new column for this using a suitable method.

In [44]:
#create a new empty list
#find the days
result2 = []

for value in df_ords_prods_merge["orders_day_of_week"]:
  if value == 0  or value == 1:
    result2.append("Busiest days")
  elif value == 4 or value == 3:
    result2.append("Slowest days")
  else:
    result2.append("Regularly busy")

In [45]:
#create a new column in df and fill it up with elements of result2 list
df_ords_prods_merge['busiest_days'] = result2

In [47]:
df_ords_prods_merge['busiest_days'].value_counts (dropna=False)

busiest_days
Regularly busy    12927873
Busiest days      11875759
Slowest days       7631438
Name: count, dtype: int64

#### 03 Check the values of this new column for accuracy. Note any observations in markdown format.
orders_day_of_week
0.0    6209808
1.0    5665951
6.0    4500391
2.0    4217868
5.0    4209603
3.0    3844175
4.0    3787263
NaN         11


busiest_days
6+2+5  Regularly busy    12927873   incorrect more with 11 because of the rows with missing values (NaN)
0+1    Busiest days      11875759   correct
3+4    Slowest days       7631438   correct

#### 04 The senior technical officer at Instacart wants you to identify the busiest hours of the day. Rather than by hour, they want periods of time labeled “Most orders,” “Average orders,” and “Fewest orders.” Create a new column containing these labels called “busiest_period_of_day.”

In [48]:
df_ords_prods_merge.columns

Index(['Unnamed: 0', 'product_id', 'product_name', 'aisle_id', 'department_id',
       'prices', 'order_id', 'user_id', 'order_number', 'orders_day_of_week',
       'order_time_of_day', 'days_since_prior_order', 'add_to_cart_order',
       'reordered', '_merge', 'price_range_loc', 'busiest_day',
       'busiest_days'],
      dtype='object')

In [49]:
#in which hours most orders take place

df_ords_prods_merge['order_time_of_day'].value_counts (dropna=False)

order_time_of_day
10.0    2764476
11.0    2738647
14.0    2691598
15.0    2664583
13.0    2663346
12.0    2620898
16.0    2537506
9.0     2456751
17.0    2089510
8.0     1719991
18.0    1637956
19.0    1259416
20.0     977049
7.0      891951
21.0     796379
22.0     634743
23.0     402621
6.0      290796
0.0      218951
1.0      115787
5.0       88064
2.0       69435
4.0       53284
3.0       51321
NaN          11
Name: count, dtype: int64

I have quickly made a chart in Excel to create the periods easily

Most orders : 10 11 14 15 13 12 16 9
Average orders 17 8 18 19 20 7 21 22
Fewest orders 23 6 0 1 5 2 4 3

In [50]:
# create the categories
#create a new empty list

result_hours = []

for value in df_ords_prods_merge["order_time_of_day"]:
  if value in [10, 11, 14, 15, 13, 12, 16, 9]:
    result_hours.append("Most orders")
  elif value in [17, 8, 18, 19, 20, 7, 21, 22]:
    result_hours.append("Average orders")
  elif value in [23, 6, 0, 1, 5, 2, 4, 3]:
    result_hours.append("Fewest orders")
  else:
    result_hours.append("Missing data")

In [51]:
#create a new column in df and fill it up with elements of result_hours list
df_ords_prods_merge['busiest_period_of_day'] = result_hours

#### 05 Print the frequency for this new column.

In [52]:
df_ords_prods_merge['busiest_period_of_day'].value_counts (dropna=False)

busiest_period_of_day
Most orders       21137805
Average orders    10006995
Fewest orders      1290259
Missing data            11
Name: count, dtype: int64

# Export pickle

In [53]:
df_ords_prods_merge.to_pickle (r'C:\Users\User\OneDrive\Dokumentumok\CF\Phyton\Instacart Basket Analysis\02 Data\Prepared Data\orders_products_new_var.pkl')